src/systems/eigen_system.C

```diff
@@ -123,158 +129,85 @@ void EigenSystem::set_eigenproblem_type (EigenProblemType ept)
+      }
+    else if (!precond_matrix)
+      precond_matrix = &(this->add_matrix(precond_matrix_name()));
+  }
+  else
+  {
+    if (!matrix_A)
+      matrix_A = &(this->add_matrix(matrix_A_name()));
+
+    if (generalized() && !matrix_B)
+      matrix_B = &(this->add_matrix(matrix_B_name()));
+  }
 }
 
 
 
 void EigenSystem::init_matrices ()
 {
-  DofMap & dof_map = this->get_dof_map();
-
-  if (_use_shell_matrices)
-  {
-    shell_matrix_A = ShellMatrix<Number>::build(this->comm());
-
-    shell_matrix_A->attach_dof_map(dof_map);
-
-    shell_matrix_A->init();
+  Parent::init_matrices();
 
-    if (_is_generalized_eigenproblem)
+  if (shell_matrix_A)
     {
-       shell_matrix_B = ShellMatrix<Number>::build(this->comm());
-
-       shell_matrix_B->attach_dof_map(dof_map);
-
-       shell_matrix_B->init();
+      shell_matrix_A->attach_dof_map(this->get_dof_map());
+      shell_matrix_A->init();
     }
 
-    if (_use_shell_precond_matrix)
+  if (shell_matrix_B)
     {
-      shell_precond_matrix = ShellMatrix<Number>::build(this->comm());
-
-      shell_precond_matrix->attach_dof_map(dof_map);
-
-      shell_precond_matrix->init();
+      shell_matrix_B->attach_dof_map(this->get_dof_map());
+      shell_matrix_B->init();
     }
-    else
-    {
-      precond_matrix = SparseMatrix<Number>::build(this->comm());
-
-      dof_map.attach_matrix(*precond_matrix);
-
-      dof_map.compute_sparsity(this->get_mesh());
 
-      precond_matrix->init();
-
-      precond_matrix->zero();
+  if (shell_precond_matrix)
+    {
+      shell_precond_matrix->attach_dof_map(this->get_dof_map());
+      shell_precond_matrix->init();
     }
-  }
-  else
-  {
-    // build the system matrix
-    matrix_A = SparseMatrix<Number>::build(this->comm());
-
-    dof_map.attach_matrix(*matrix_A);
-
-    // build matrix_B only in case of a
-    // generalized problem
-    if (_is_generalized_eigenproblem)
-      {
-        matrix_B = SparseMatrix<Number>::build(this->comm());
-        dof_map.attach_matrix(*matrix_B);
-      }
-
-    dof_map.compute_sparsity(this->get_mesh());
-
-    // initialize and zero system matrix
-    matrix_A->init();
-    matrix_A->zero();
-
-    // eventually initialize and zero system matrix_B
-    if (_is_generalized_eigenproblem)
-      {
-        matrix_B->init();
-        matrix_B->zero();
-      }
-  }
 }
 
 
-
 void EigenSystem::reinit ()
 {
   // initialize parent data
+  // this calls reinit on matrix_A, matrix_B, and precond_matrix (if any)
   Parent::reinit();
 
-  DofMap & dof_map = this->get_dof_map();
-
-  if (_use_shell_matrices)
-  {
-    shell_matrix_A->clear();
-
-    if (_is_generalized_eigenproblem)
-     shell_matrix_B->clear();
-
-    if (_use_shell_precond_matrix)
-      shell_precond_matrix->clear();
-    else
+  if (shell_matrix_A)
     {
-      precond_matrix->clear();
-      dof_map.clear_sparsity();
-      dof_map.compute_sparsity(this->get_mesh());
+      shell_matrix_A->clear();
+      shell_matrix_A->init();
```

Comment:
Should we reattach `dof_map`?

Reply:
Calling `clear()` doesn't remove the dof map, so it's not needed